core/src/main/java/jenkins/security/ClassFilterImpl.java

```diff
@@ -0,0 +1,259 @@
+                        if (isPluginManifest(mf)) {
+                            LOGGER.log(Level.FINE, "{0} seems to be a Jenkins plugin, OK", loc);
+                            return true;
+                        } else {
+                            LOGGER.log(Level.FINE, "{0} does not look like a Jenkins plugin", loc);
+                        }
+                    } else {
+                        LOGGER.log(Level.FINE, "ignoring {0} with no manifest", loc);
+                    }
+                } catch (Exception x) {
+                    LOGGER.log(Level.WARNING, "problem checking " + loc, x);
+                }
+            }
+            Matcher m = CLASSES_JAR.matcher(loc);
+            if (m.matches()) {
+                // Cf. ClassicPluginStrategy.createClassJarFromWebInfClasses: handle legacy plugin format with unpacked WEB-INF/classes/
+                try {
+                    File manifestFile = new File(new URI(m.group(1) + "META-INF/MANIFEST.MF"));
+                    if (manifestFile.isFile()) {
+                        try (InputStream is = new FileInputStream(manifestFile)) {
+                            if (isPluginManifest(new Manifest(is))) {
+                                LOGGER.log(Level.FINE, "{0} looks like a Jenkins plugin based on {1}, OK", new Object[] {loc, manifestFile});
+                                return true;
+                            } else {
+                                LOGGER.log(Level.FINE, "{0} does not look like a Jenkins plugin", manifestFile);
+                            }
+                        }
+                    } else {
+                        LOGGER.log(Level.FINE, "{0} has no matching {1}", new Object[] {loc, manifestFile});
+                    }
+                } catch (Exception x) {
+                    LOGGER.log(Level.WARNING, "problem checking " + loc, x);
+                }
+            }
+            if (Main.isUnitTest) {
+                if (loc.endsWith("/target/classes/")) {
+                    LOGGER.log(Level.FINE, "{0} seems to be current plugin classes, OK", loc);
+                    return true;
+                }
+                if (loc.endsWith("/target/test-classes/") || loc.endsWith("-tests.jar")) {
+                    LOGGER.log(Level.FINE, "{0} seems to be test classes, OK", loc);
+                    return true;
+                }
+                if (loc.matches(".+/jenkins-test-harness-.+[.]jar")) {
+                    LOGGER.log(Level.FINE, "{0} seems to be jenkins-test-harness, OK", loc);
+                    return true;
+                }
+            }
+            LOGGER.log(Level.FINE, "{0} is not recognized; rejecting", loc);
+            return false;
+        });
+    }
+
+    private static boolean isPluginManifest(Manifest mf) {
+        Attributes attr = mf.getMainAttributes();
+        return attr.getValue("Short-Name") != null && (attr.getValue("Plugin-Version") != null || attr.getValue("Jenkins-Version") != null) ||
```

Comment:
So it may not work for obsolete Hudson plugins, which will need an update. Fine for me

Reply:
They should have `Plugin-Version` at least. E.g. from `maven-plugin-1.320.jar!/META-INF/MANIFEST.MF`:

```
…
Short-Name: maven-plugin
Long-Name: Maven Integration plugin
Url: https://hudson.dev.java.net/pom/maven-plugin
Plugin-Version: 1.320
Hudson-Version: 1.320
```